README.md

```diff
@@ -0,0 +1,138 @@
+* `Controller.prototype.generateMultipartRequest(opts)`: generate a list of
+  standard interchange format objects which can be run to upload the file.
+  The `bucket`, `key` and `uploadId` parameters must be included.  The `parts`
+  parameter is a list of objects like `{sha256: '<sha256>', size: 1234}`.
+  Note that the ordering of the list implies the ordering of the parts.  This
+  method returns a list of signed requests which should be run by the
+  untrusted host to upload the file.  Note that the order of these requests
+  is equivalent by the order of the `parts` list
+* `Controller.prototype.completeMultipartUpload(opts)`: instruct S3 to commit
+  the upload to S3.  This will make the file available immediately.  The `opts`
+  parameters `bucket`, `key` and `uploadId` parameters must be provided. The 
+  `etags` parameter is a list of etags returned by the S3 API for the upload
+  of each part.  This method returns the `etag` of the completed upload.
+* `Controller.prototype.abortMultipartUpload(opts)`: instruct S3 to abort
+  the upload.  The `bucket`, `key` and `uploadId` parameters must be provided.
+  This method returns nothing
+* `Controller.prototype.generateSinglepartRequest(opts)`: Generate a single
+  request in the standard interchange format to upload an object to S3 and
+  return it.  The `bucket`, `key`, `sha256` and `size` parameters must be
+  specified
+
+
+There is a `Runner` class which is used by the `Controller` to run the requests
+which occur on the server.  This same `Runner` class is used to run requests on
+the untrusted host.  This is done to share code as well as ensure that the code
+which is published for consumers is working.
+
+The interchange format for the `Runner` class is like so:
+
+```json
+{
+  "url": "https://www.hostname.com:443/path/to/resource?query=string",
+  "method": "GET",
+  "headers": {
+    "Content-Length": "1234",
+  }
+}
+```
+
+Note that the request body itself is not specified anywhere in this format.
+
+The `Runner` class has the following methods:
+
+* `new Runner(agent, agentOpts)`: create a new `Runner` instance.  The `agent`
+  parameter is the `https.Agent` to use for all requests.  If the `agent`
+  parameter is falsy, then the `agentOpts` parameter will be passed to the
+  constructor like so `new https.Agent(agentOpts)`
+* `Runner.prototype.run(request, body, noThrow)`: Run a request using a standard
+  interchange format object `request`.  A `string`, `Buffer` or `Readable`
+  stream as the `body` parameter will be used to determine obtain the request
+  body.  Note that for `string` and `Buffer` bodies, the entire body must
+  be stored in memory.  The `Readable` stream option is recommended for all
+  but the smallest files
+
+The `Runner.prototype.run` method returns an object which looks like this:
+```javascript
+{
+  body: "<entire response body>",
+  headers: {
+    aresponse: "header"
+  },
+  statusCode: 200,
```

Comment:
Except that for downloading (which will be implemented in this library shortly) will require streaming bodies.  It originally did what you said, but I realised that I needed download support.